dataset/src/main/spark-3.4+/frameless/MapGroups.scala

```diff
@@ -1,21 +0,0 @@
-package frameless
-
-import org.apache.spark.sql.Encoder
-import org.apache.spark.sql.catalyst.expressions.Attribute
-import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, MapGroups => SMapGroups}
-
-object MapGroups {
```

Comment:
Q. What makes it impossible to keep these MapGroups here, and use the override from the shim in the client code?

The client code will pick up the version from shim and it will enforce MapGroups implementation from the dependency.

We can play around with it thats for sure, but I am curious how we can get it as far as possible to the user side not enforcing behavior / shims usage by the lib implementation.

Reply:
^ Also given how convoluted https://github.com/sparkutils/shim looks, id really like to find a solution so user replaces / uses their own implementation of the critical classes (i.e. by pulling shims into deps)

Reply:
> ^ Also given how convoluted https://github.com/sparkutils/shim looks, id really like to find a solution so user replaces / uses their own implementation of the critical classes (i.e. by pulling shims into deps)

that's the point of the shim lib.  When a user wants an upgrade (and the shim is there) they use the newer shim, no need to change the frameless version used.  The complexity is there, it's just about how to manage it, I'd personally prefer it is in a single lib rather than in client code.  Currently only the testing is version specific.

You could, per the other point, move the call to impl out into a typeclass , this wouldn't impact the users source all that much if a default impl passing to shim existed. (a recompile would be needed of course for the extra implicit).  That could give the best of both worlds - point location fixes and a bunch of default fixes.